Pre-built binaries not installable for ... and node@8.11.1 (node-v57 ABI, unknown) (falling back to source compile with node-gyp)

Hi guys,

  since yesterday I cannot retrieve my pre-built binaries (posted on amazon S3) using "latest" node-pre-gyp.

I suspect a regression in node-pre-gyp..

Here is what I get:

```
Pre-built binaries not installable for xxxxx and node@8.11.1 (node-v57 ABI, unknown) (falling back to source compile with node-gyp)
```

Thanks to investigate asap, this is a blocking issue.

I am having a similar issue with bcrypt. 

```
> bcrypt@2.0.0 install /tmp/node/node_modules/bcrypt
> node-pre-gyp install --fallback-to-build

node-pre-gyp ERR! Pre-built binaries not installable for bcrypt@2.0.0 and node@9.11.1 (node-v59 ABI, glibc) (falling back to source compile with node-gyp) 
node-pre-gyp ERR! Hit error EEXIST: file already exists, mkdir '/tmp/node/node_modules/bcrypt/lib/binding' 
make: Entering directory '/tmp/node/node_modules/bcrypt/build'
  CXX(target) Release/obj.target/bcrypt_lib/src/blowfish.o
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt.o
../src/bcrypt.cc: In function ‘void encode_salt(char*, u_int8_t*, char, u_int16_t, u_int8_t)’:
../src/bcrypt.cc:114:1: warning: ‘__builtin___snprintf_chk’ output may be truncated before the last format character [-Wformat-truncation=]
 encode_salt(char *salt, u_int8_t *csalt, char minor, u_int16_t clen, u_int8_t logr)
 ^~~~~~~~~~~
In file included from /usr/include/stdio.h:862:0,
                 from ../src/bcrypt.cc:34:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:65:44: note: ‘__builtin___snprintf_chk’ output between 4 and 5 bytes into a destination of size 4
        __bos (__s), __fmt, __va_arg_pack ());
                                            ^
../src/bcrypt.cc: In function ‘void bcrypt(const char*, const char*, char*)’:
../src/bcrypt.cc:147:1: warning: ‘__builtin___snprintf_chk’ output may be truncated before the last format character [-Wformat-truncation=]
 bcrypt(const char *key, const char *salt, char *encrypted)
 ^~~~~~
In file included from /usr/include/stdio.h:862:0,
                 from ../src/bcrypt.cc:34:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:65:44: note: ‘__builtin___snprintf_chk’ output between 4 and 5 bytes into a destination of size 4
        __bos (__s), __fmt, __va_arg_pack ());
                                            ^
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt_node.o
  SOLINK_MODULE(target) Release/obj.target/bcrypt_lib.node
  COPY Release/bcrypt_lib.node
  COPY /tmp/node/node_modules/bcrypt/lib/binding/bcrypt_lib.node
mkdir: cannot create directory ‘/tmp/node/node_modules/bcrypt/lib/binding/’: File exists
action_after_build.target.mk:8: recipe for target '/tmp/node/node_modules/bcrypt/lib/binding/bcrypt_lib.node' failed
make: *** [/tmp/node/node_modules/bcrypt/lib/binding/bcrypt_lib.node] Error 1
make: Leaving directory '/tmp/node/node_modules/bcrypt/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:258:23)
gyp ERR! stack     at ChildProcess.emit (events.js:180:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:209:12)
gyp ERR! System Linux 4.13.0-38-generic
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "build" "--fallback-to-build" "--module=/tmp/node/node_modules/bcrypt/lib/binding/bcrypt_lib.node" "--module_name=bcrypt_lib" "--module_path=/tmp/node/node_modules/bcrypt/lib/binding"
gyp ERR! cwd /tmp/node/node_modules/bcrypt
gyp ERR! node -v v9.11.1
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok 
node-pre-gyp ERR! build error 
node-pre-gyp ERR! stack Error: Failed to execute '/usr/bin/node /usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/tmp/node/node_modules/bcrypt/lib/binding/bcrypt_lib.node --module_name=bcrypt_lib --module_path=/tmp/node/node_modules/bcrypt/lib/binding' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/tmp/node/node_modules/node-pre-gyp/lib/util/compile.js:83:29)
node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:180:13)
node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:936:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:220:5)
node-pre-gyp ERR! System Linux 4.13.0-38-generic
node-pre-gyp ERR! command "/usr/bin/node" "/tmp/node/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd /tmp/node/node_modules/bcrypt
node-pre-gyp ERR! node -v v9.11.1
node-pre-gyp ERR! node-pre-gyp -v v0.7.0
node-pre-gyp ERR! not ok 
Failed to execute '/usr/bin/node /usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/tmp/node/node_modules/bcrypt/lib/binding/bcrypt_lib.node --module_name=bcrypt_lib --module_path=/tmp/node/node_modules/bcrypt/lib/binding' (1)
npm WARN node@1.0.0 No description
npm WARN node@1.0.0 No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bcrypt@2.0.0 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the bcrypt@2.0.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/amitosh/.npm/_logs/2018-04-10T18_17_52_456Z-debug.log
```

It seems that the folder is not being created, rather a file with contents any random file from the download tarball is being added.

--- a/lib/package.js
+++ b/lib/package.js
@@ -33,15 +33,18 @@ function _package(gyp, argv, callback) {
             return true;
         };
         mkdirp(path.dirname(tarball),function(err) {
-            from = path.dirname(from);
             if (err) return callback(err);
             packlist({ path: from }).then(function(files) {
+                var base = path.basename(from);
+                files = files.map(function(file) {
+                    return path.join(base, file);
+                });
                 tar.create({
                     portable: true,
                     gzip: true,
                     onentry: filter_func,
                     file: tarball,
-                    cwd: from
+                    cwd: path.dirname(from)
                 }, files, function(err) {
                     if (err)  console.error('['+package_json.name+'] ' + err.message);
                     else log.info('package','Binary staged at "' + tarball + '"');

--- a/test/app8/.gitignore
+++ b/test/app8/.gitignore
@@ -1,0 +1,5 @@
+.DS_Store
+build/
+lib/
+node_modules
+npm-debug.log

--- a/test/app8/README.md
+++ b/test/app8/README.md
@@ -1,0 +1,3 @@
+# Test app
+
+Demonstrates a simpler configuration that uses node-pre-gyp.

--- a/test/app8/app8.cc
+++ b/test/app8/app8.cc
@@ -1,0 +1,38 @@
+// v8
+#include <v8.h>
+
+// node.js
+#include <node.h>
+#include <node_version.h>
+
+#if (NODE_MODULE_VERSION > 0x000B)
+
+    static void get_hello(const v8::FunctionCallbackInfo<v8::Value>& args)
+    {
+        v8::HandleScope scope(v8::Isolate::GetCurrent());
+        args.GetReturnValue().Set(v8::String::NewFromUtf8(v8::Isolate::GetCurrent(),"hello"));
+    }
+
+#else
+
+    static v8::Handle<v8::Value> get_hello(const v8::Arguments& args)
+    {
+        v8::HandleScope scope;
+        return scope.Close(v8::String::New("hello"));
+    }
+
+#endif
+
+extern "C" {
+    static void start(v8::Handle<v8::Object> target) {
+#if (NODE_MODULE_VERSION > 0x000B)
+        v8::HandleScope scope(v8::Isolate::GetCurrent());
+#else
+        v8::HandleScope scope;
+#endif
+        NODE_SET_METHOD(target, "hello", get_hello);
+    }
+}
+
+NODE_MODULE(app8, start)
+

--- a/test/app8/binding.gyp
+++ b/test/app8/binding.gyp
@@ -1,0 +1,13 @@
+{
+  "targets": [
+    {
+      "target_name": "<(module_name)",
+      "sources": [ "<(module_name).cc" ],
+      "product_dir": "<(module_path)",
+      "xcode_settings": {
+        "MACOSX_DEPLOYMENT_TARGET":"10.9",
+        "CLANG_CXX_LIBRARY": "libc++"
+      }
+     }
+  ]
+}

--- a/test/app8/index.js
+++ b/test/app8/index.js
@@ -1,0 +1,6 @@
+var binary = require('node-pre-gyp');
+var path = require('path')
+var binding_path = binary.find(path.resolve(path.join(__dirname,'./package.json')));
+var binding = require(binding_path);
+
+require('assert').equal(binding.hello(),"hello");

--- a/test/app8/package.json
+++ b/test/app8/package.json
@@ -1,0 +1,21 @@
+{
+    "name": "node-pre-gyp-test-app8",
+    "author": "Dane Springmeyer <springmeyer>",
+    "description": "node-pre-gyp test",
+    "repository": {
+        "type": "git",
+        "url": "git://github.com/mapbox/node-pre-gyp.git"
+    },
+    "license": "BSD-3-Clause",
+    "version": "0.1.0",
+    "main": "./index.js",
+    "binary": {
+        "module_name": "app8",
+        "module_path": "./lib",
+        "host": "https://node-pre-gyp-tests.s3-us-west-1.amazonaws.com"
+    },
+    "scripts": {
+        "install": "node-pre-gyp install --fallback-to-build",
+        "test": "node index.js"
+    }
+}

--- a/test/build.test.js
+++ b/test/build.test.js
@@ -12,6 +12,8 @@ var versioning = require('../lib/util/versioning.js');
 var tar = require('tar');
 
 var localVer = [versioning.get_runtime_abi('node'), process.platform, process.arch].join('-');
+var SOEXT = process.platform === 'darwin' ? 'dylib': 'so';
+
 // The list of different sample apps that we use to test
 var apps = [
     {
@@ -37,7 +39,7 @@ var apps = [
     {
         'name': 'app4',
         'args': '',
-        'files': [path.join(localVer, 'app4.node'), path.join(localVer, 'lib.target', 'mylib.dylib')]
+        'files': [path.join(localVer, 'app4.node'), path.join(localVer, 'lib.target', 'mylib.' + SOEXT)]
     },
     {
         'name': 'app7',

--- a/test/build.test.js
+++ b/test/build.test.js
@@ -8,32 +8,43 @@ var rm = require('rimraf');
 var path = require('path');
 var getPrevious = require('./target_version.util.js');
 var napi = require ('../lib/util/napi.js');
+var tar = require('tar');
 
 // The list of different sample apps that we use to test
 var apps = [
     {
         'name': 'app1',
-        'args': ''
-    },
-    {
-         'name': 'app2',
-         'args': '--custom_include_path=../include --debug'
+        'args': '',
+        'files': ['binding/app1.node']
     },
     {
         'name': 'app2',
-        'args': '--custom_include_path=../include --toolset=cpp11'
+        'args': '--custom_include_path=../include --debug',
+        'files': ['node-pre-gyp-test-app2/app2.node']
+    },
+    {
+        'name': 'app2',
+        'args': '--custom_include_path=../include --toolset=cpp11',
+        'files': ['node-pre-gyp-test-app2/app2.node']
     },
     {
         'name': 'app3',
-        'args': ''
+        'args': '',
+        'files': ['node-v48-darwin-x64/app3.node']
     },
     {
         'name': 'app4',
-        'args': ''
+        'args': '',
+        'files': ['node-v48-darwin-x64/app4.node', 'node-v48-darwin-x64/lib.target/mylib.dylib']
     },
     {
         'name': 'app7',
         'args': ''
+    },
+    {
+        'name': 'app8',
+        'args': '',
+        'files': ['lib/app8.node']
     }
 ];
 
@@ -211,7 +222,30 @@ apps.forEach(function(app) {
         test(app.name + ' packages ' + app.args, function(t) {
             run('node-pre-gyp', 'package', '', app, {}, function(err,stdout,stderr) {
                 t.ifError(err);
-                t.end();
+                // Make sure a tarball was created
+                run('node-pre-gyp', 'reveal', 'staged_tarball --silent', app, {}, function(err,stdout,stderr) {
+                    t.ifError(err);
+                    var staged_tarball = stdout.trim();
+                    if (staged_tarball.indexOf('\n') !== -1) { // take just the first line
+                        staged_tarball = staged_tarball.substr(0,staged_tarball.indexOf('\n'));
+                    }
+                    var tarball_path = path.join(__dirname, app.name, staged_tarball);
+                    t.ok(existsSync(tarball_path),'staged tarball is a valid file');
+                    if (!app.files) {
+                        return t.end();
+                    }
+                    // Make sure the package contains what we expect
+                    var entries = [];
+                    tar.t({
+                        file: tarball_path,
+                        sync: true,
+                        onentry: function (entry) {
+                            entries.push(entry.path);
+                        }
+                    });
+                    t.same(entries.sort(), app.files.sort(), 'staged tarball contains the right files');
+                    t.end();
+                });
             });
         });
 

--- a/test/build.test.js
+++ b/test/build.test.js
@@ -8,8 +8,10 @@ var rm = require('rimraf');
 var path = require('path');
 var getPrevious = require('./target_version.util.js');
 var napi = require ('../lib/util/napi.js');
+var versioning = require('../lib/util/versioning.js');
 var tar = require('tar');
 
+var localVer = [versioning.get_runtime_abi('node'), process.platform, process.arch].join('-');
 // The list of different sample apps that we use to test
 var apps = [
     {
@@ -30,12 +32,12 @@ var apps = [
     {
         'name': 'app3',
         'args': '',
-        'files': ['node-v48-darwin-x64/app3.node']
+        'files': [path.join(localVer, 'app3.node')]
     },
     {
         'name': 'app4',
         'args': '',
-        'files': ['node-v48-darwin-x64/app4.node', 'node-v48-darwin-x64/lib.target/mylib.dylib']
+        'files': [path.join(localVer, 'app4.node'), path.join(localVer, 'lib.target', 'mylib.dylib')]
     },
     {
         'name': 'app7',

--- a/test/build.test.js
+++ b/test/build.test.js
@@ -12,34 +12,100 @@ var versioning = require('../lib/util/versioning.js');
 var tar = require('tar');
 
 var localVer = [versioning.get_runtime_abi('node'), process.platform, process.arch].join('-');
-var SOEXT = process.platform === 'darwin' ? 'dylib': 'so';
+var SOEXT = {'darwin': 'dylib', 'linux': 'so', 'win32': 'dll'}[process.platform];
 
 // The list of different sample apps that we use to test
 var apps = [
     {
         'name': 'app1',
         'args': '',
-        'files': ['binding/app1.node']
+        'files': {
+            'base': ['binding/app1.node'],
+            'win32': {
+                'base': [
+                    'binding/app1.exp',
+                    'binding/app1.lib',
+                    'binding/app1.map',
+                    'binding/app1.node'
+                ],
+                'node-v59': [
+                    'binding/app1.exp',
+                    'binding/app1.iobj',
+                    'binding/app1.ipdb',
+                    'binding/app1.lib',
+                    'binding/app1.map',
+                    'binding/app1.node'
+                ]
+            }
+        }
     },
     {
         'name': 'app2',
         'args': '--custom_include_path=../include --debug',
-        'files': ['node-pre-gyp-test-app2/app2.node']
+        'files': {
+            'base': ['node-pre-gyp-test-app2/app2.node'],
+            'win32': [
+                'node-pre-gyp-test-app2/app2.exp',
+                'node-pre-gyp-test-app2/app2.ilk',
+                'node-pre-gyp-test-app2/app2.lib',
+                'node-pre-gyp-test-app2/app2.map',
+                'node-pre-gyp-test-app2/app2.node'
+            ]
+        }
     },
     {
         'name': 'app2',
         'args': '--custom_include_path=../include --toolset=cpp11',
-        'files': ['node-pre-gyp-test-app2/app2.node']
+        'files': {
+            'base': ['node-pre-gyp-test-app2/app2.node'],
+            'win32': {
+                'base': [
+                    'node-pre-gyp-test-app2/app2.exp',
+                    'node-pre-gyp-test-app2/app2.lib',
+                    'node-pre-gyp-test-app2/app2.map',
+                    'node-pre-gyp-test-app2/app2.node'
+                ],
+                'node-v59': [
+                    'node-pre-gyp-test-app2/app2.exp',
+                    'node-pre-gyp-test-app2/app2.iobj',
+                    'node-pre-gyp-test-app2/app2.ipdb',
+                    'node-pre-gyp-test-app2/app2.lib',
+                    'node-pre-gyp-test-app2/app2.map',
+                    'node-pre-gyp-test-app2/app2.node' 
+                ]
+            }
+        }
     },
     {
         'name': 'app3',
         'args': '',
-        'files': [path.join(localVer, 'app3.node')]
+        'files': {
+            'base': [[localVer, 'app3.node'].join('/')],
+            'win32': {
+                'base': [
+                    [localVer, 'app3.exp'].join('/'),
+                    [localVer, 'app3.lib'].join('/'),
+                    [localVer, 'app3.map'].join('/'),
+                    [localVer, 'app3.node'].join('/')
+                ],
+                'node-v59': [
+                    [localVer, 'app3.exp'].join('/'),
+                    [localVer, 'app3.iobj'].join('/'),
+                    [localVer, 'app3.ipdb'].join('/'),
+                    [localVer, 'app3.lib'].join('/'),
+                    [localVer, 'app3.map'].join('/'),
+                    [localVer, 'app3.node'].join('/'),
+                ]
+            }
+        }
     },
     {
         'name': 'app4',
         'args': '',
-        'files': [path.join(localVer, 'app4.node'), path.join(localVer, 'lib.target', 'mylib.' + SOEXT)]
+        'files': {
+            'base': [[localVer, 'app4.node'].join('/'), [localVer, 'lib.target', 'mylib.' + SOEXT].join('/')],
+            'win32': [[localVer, 'app4.node'].join('/'), [localVer, 'mylib.' + SOEXT].join('/')]
+        }
     },
     {
         'name': 'app7',
@@ -48,7 +114,25 @@ var apps = [
     {
         'name': 'app8',
         'args': '',
-        'files': ['lib/app8.node']
+        'files': {
+            'base': ['lib/app8.node'],
+            'win32': {
+                'base': [
+                    'lib/app8.exp',
+                    'lib/app8.lib',
+                    'lib/app8.map',
+                    'lib/app8.node'
+                ],
+                'node-v59': [
+                    'lib/app8.exp',
+                    'lib/app8.iobj',
+                    'lib/app8.ipdb',
+                    'lib/app8.lib',
+                    'lib/app8.map',
+                    'lib/app8.node'
+                ]
+            }
+        }
     }
 ];
 
@@ -247,7 +331,19 @@ apps.forEach(function(app) {
                             entries.push(entry.path);
                         }
                     });
-                    t.same(entries.sort(), app.files.sort(), 'staged tarball contains the right files');
+                    var files = app.files.base;
+                    var nodever = versioning.get_runtime_abi('node');
+                    // Look for a more specific choice
+                    if (app.files.hasOwnProperty(process.platform)) {
+                        if (app.files[process.platform].hasOwnProperty(nodever)) {
+                            files = app.files[process.platform][nodever];
+                        } else if (app.files[process.platform].hasOwnProperty('base')) {
+                            files = app.files[process.platform].base; 
+                        } else {
+                            files = app.files[process.platform];
+                        }
+                    }
+                    t.same(entries.sort(), files.sort(), 'staged tarball contains the right files');
                     t.end();
                 });
             });

